### PR TITLE
gzip and bzip2 stream

### LINKED ISSUE #610: compression / decompression streams
Add compression/decompression streams
- [ ] gzip
- [ ] bzip2
- [ ] bgzf

* code
* tests
  * viability of type erasure through `std::ostream *`
* benchmarks
  * against each other
  * against seqan2
  * type-erased VS not type-erased

### PR DESCRIPTION
**Pull request by h-2**
resolves #610 

no bgzf for now, because it has additional requirements
not tied into any of the files, yet

Benchmarks have high variance on my system, but here is one:

### ostream
```
-------------------------------------------------------------------------------------
Benchmark                                              Time           CPU Iterations
-------------------------------------------------------------------------------------
uncompressed                                           6 ns          6 ns  144466035
compressed<contrib::gz_ostream>                        8 ns          8 ns   90665658
compressed<contrib::bz2_ostream>                     529 ns        520 ns    1807077
compressed_type_erased<contrib::gz_ostream>            8 ns          8 ns   92377217
compressed_type_erased<contrib::bz2_ostream>         583 ns        576 ns    1000000
compressed_type_erased2<contrib::gz_ostream>           8 ns          8 ns   91140686
compressed_type_erased2<contrib::bz2_ostream>        405 ns        400 ns    2322257
seqan2_compressed<seqan::Nothing>                     14 ns         14 ns   59982816
seqan2_compressed<seqan::GZFile>                      24 ns         23 ns   51139189
seqan2_compressed<seqan::BgzfFile>                     9 ns          9 ns   78701854
seqan2_compressed<seqan::BZ2File>                    568 ns        561 ns    1889330

```

Take-home:
* doing type-erasure the way I proposed is basically free
* seqan2's virtual streams incur a noticeable penalty
* bzip2 is soo slow

### istream 
```
------------------------------------------------------------------------------------
Benchmark                                             Time           CPU Iterations
------------------------------------------------------------------------------------
uncompressed                                          2 ns          2 ns  341142583
compressed<contrib::gz_istream>                       3 ns          3 ns  207857710
compressed_type_erased<contrib::gz_istream>           3 ns          3 ns  207806119
compressed_type_erased2<contrib::gz_istream>          4 ns          4 ns  170516471
seqan2_compressed<seqan::Nothing>                     6 ns          6 ns  119800598
seqan2_compressed<seqan::GZFile>                     10 ns         10 ns   92157975
```
Same thing here (BZip2 bench not run by default, because the setup is so expensive – need to build a a sufficiently large buffer to read from and compression is very slow compared with decompression)

### CODE REVIEW

**Comment by h-2** [comment]
@marehr I tried to follow the mantra of one test file per header file, although it adds some redundancy....

**Comment by marehr** [comment]
> @marehr I tried to follow the mantra of one test file per header file, although it adds some redundancy....

Thank you!

**Comment by marehr** [comment]
@h-2 Can we add as first commit the original code in contrib, s.t. we can see which changes you did?

**Comment by h-2** [comment]
> @h-2 Can we add as first commit the original code in contrib, s.t. we can see which changes you did?

I don't think it would be very helpful, because it lot was moved around, and you wouldn't actually see the individual things. Also lots of diff-noise.

Structural changes (these would be hard to follow):
* Originally there was one file with the declarations and one file with most of the definitions (but not all...); this might have been header and cpp once, but in SeqAn2 it was header and header_impl
* this distinction didn't make much sense as you needed both anyway.
* on the other hand there was no distinction between input and output on file level, something that does make a lot of sense from my POV, because often you only use one or the other.
* I changed this so that we have different files for in VS out, but not for decl VS def

Minor changes:
* some things were called "zip" vs "unzip" instead of "output" vs "input", I changed that to be more consistent.
* there was also inconsistent use of the terms "zip", "zlib", "gzip", "gz"... I just changed everything to gz and bz2 respectively so that we always have EXTENSION_ostream and EXTENSION_istream now
* added `#pragma once`
* moved stuff to namespace `seqan3::contrib` to ensure that it doesn't conflict with a potential second inclusion of the original.
* added check for our macro and print error if not found

File: include/seqan3/contrib/stream/bz2_ostream.hpp [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+#include <bzlib.h>
+
+namespace seqan3::contrib
+{
+
+// --------------------------------------------------------------------------
+// Class basic_bz2_ostreambuf
+// --------------------------------------------------------------------------
+
+const size_t BZ2_OUTPUT_DEFAULT_BUFFER_SIZE = 4096;
```
[the remark this answers is not in the document]
Comment (h-2) [reply include/seqan3/contrib/stream/bz2_ostream.hpp]: I just tried it, makes absolutely no difference on the benchmark.

[the remark this answers is not in the document]
Comment (rrahn) [reply include/seqan3/contrib/stream/bz2_ostream.hpp]: Ok, then we can leave it as is.
